shell/common/rasterizer.cc

```diff
@@ -436,6 +437,47 @@ static sk_sp<SkData> ScreenshotLayerTreeAsImage(
   return SkData::MakeWithCopy(pixmap.addr32(), pixmap.computeByteSize());
 }
 
+sk_sp<SkImage> Rasterizer::MakeImageSnapshot(
+    sk_sp<SkSurface> snapshot_surface) {
+  std::unique_ptr<RendererContextSwitchManager::RendererContextSwitch>
+      context_switch = surface_->MakeRenderContextCurrent();
+  if (!context_switch->GetSwitchResult()) {
+    return nullptr;
+  }
+  auto potentially_gpu_snapshot = snapshot_surface->makeImageSnapshot();
+  if (!potentially_gpu_snapshot) {
+    FML_LOG(ERROR) << "Screenshot: unable to make image screenshot";
+    return nullptr;
+  }
+  return potentially_gpu_snapshot;
+}
+
+sk_sp<SkImage> Rasterizer::MakeRasterImage(
+    sk_sp<SkImage> potentially_gpu_snapshot) {
+  std::unique_ptr<RendererContextSwitchManager::RendererContextSwitch>
+      context_switch = surface_->MakeRenderContextCurrent();
+  if (!context_switch->GetSwitchResult()) {
+    return nullptr;
+  }
+  auto cpu_snapshot = potentially_gpu_snapshot->makeRasterImage();
+  if (!cpu_snapshot) {
+    FML_LOG(ERROR) << "Screenshot: unable to make raster image";
+    return nullptr;
+  }
+  return cpu_snapshot;
+}
+
+void Rasterizer::ScreenshotFlushCanvas(SkCanvas& canvas) {
+  std::unique_ptr<RendererContextSwitchManager::RendererContextSwitch>
+      context_switch = surface_->MakeRenderContextCurrent();
+  if (!context_switch->GetSwitchResult()) {
+    FML_LOG(ERROR)
+        << "Screenshot: unable to switch gl context to flutter's context";
+    return;
+  }
+  canvas.flush();
+}
+
```

Comment:
Refactored out these methods so we can set the gl context for each of them. @chinmaygarde